Update hazelcast enterprise

leszko opened the pull request:
Apply Hazelcast Enterprise Helm Chart for the Cloud Pak certification.

mrgadgil commented:
@leszko Thank you for the PR.
I received Hazelcast Enterprise License key yesterday evening so I can go ahead and test chart install. Thank you for the quick turn around.

The chart readme mentions that the chart can run with `ibm-restricted-psp`. But I am running into issues doing so. 
mancenter-deployment has a root user. https://github.com/leszko/ibm-charts/blob/update-hazelcast-enterprise-for-cloud-pak/community/hazelcast-enterprise/templates/mancenter-deployment.yaml#L40
Error on deployment:
```
  Warning  FailedCreate  1m (x19 over 11m)  replicaset-controller  Error creating: pods "hazel-test-hazelcast-enterprise-mancenter-744c84f6f7-" is forbidden: unable to validate against any pod security policy: [spec.containers[0].securityContext.runAsUser: Invalid value: 0: running with the root UID is forbidden]
```
Statefulset also does not contain security context with a non root user.
```
Warning  Failed     1m (x12 over 4m)  kubelet, 9.30.251.16  Error: container has runAsNonRoot and image will run as root
```
My namespace is set up `ibm-restricted-psp`.  Am I missing something here? 
I am yet to review the PR. But running as non-root is a mandatory requirement for becoming a cloud pak. So wanted to bring it to your notice.

hasancelik commented:
Hi @mrgadgil, 

Before @leszko sends a PR, I have tested the helm chart at our ICP environment(3.1.0) successfully so I have not encounter any of above issues. Here are steps which I applied, they might help you:
```
kubectl create namespace restricted
kubectl apply -f ibm-restricted-psp.yaml
kubectl apply -f ibm-restricted-cr.yaml
kubectl create rolebinding ibm-restricted-clusterrolebinding --clusterrole=ibm-restricted-clusterrole --group=system:serviceaccounts:restricted
```
I have downloaded yamls from https://github.com/IBM/cloud-pak/tree/master/spec/security/psp

When I run `getPSPs.sh` script to see all PodSecurityPolicies that are bound to a `restricted` namespace, I am getting this output:
```
Hasans-MacBook:~ hasancelik$ ./getPSPs.sh restricted
Checking PSP configuration for namespace: restricted
default (*, default)
ibm-restricted-psp (*, default)
```
I choose  `restricted` namespace from **Target Namespace** dropdown, when I am creating helm release at Catalog:
<img width="1875" alt="Screen Shot 2019-05-30 at 13 46 42" src="https://user-images.githubusercontent.com/6005622/58645949-00a2e600-830d-11e9-999f-c6dfaffe3846.png">

bgower-ibm commented:
@hasancelik :  I see the issue; ICP did not enforce PSP by default until version 3.1.1:
https://www.ibm.com/support/knowledgecenter/en/SSBS6K_3.1.1/getting_started/whats_new.html#security

I would suggest you move up to at least the latest version (at this exact moment 3.1.2)  However, ICP v3.2.0 has been announced and will be available tomorrow (May 31).

mrgadgil commented:
@hasancelik Thanks @bgower-ibm  
My ICP environment is `3.1.2 Enterprise Edition`

hasancelik commented:
Thanks @bgower-ibm for the info, as a result we need a create new ICP environment.

mrgadgil commented:
Below are my review comments. Apologize for the long comment. 
I have compared the chart with guidelines provided for charts to become cloud paks. The guidelines mostly align with helm and kubernetes best practices and some common ICP standards for charts. Please feel free to reach out to me with questions.

[Recommended] The chart uses old style of labels. It is recommended to update to update to the new format.
Refer: https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels
https://github.com/helm/helm/blob/master/docs/chart_best_practices/labels.md

[Required] Pod Security Policy guidelines
IBM recommends charts to define a custom PSP and provide details in the Readme on how to generate a custom PSP.
IBM pre-defined PSPs can change between ICP releases. The pre-defined policies grant more privileges than the chart requires to install.
We want the charts to provide guidelines so that the users can have a cluster admin create a custom PSP and narrow down the privileges to only the ones required by the chart.
Example: https://github.com/IBM/charts/tree/master/stable/ibm-websphere-liberty#podsecuritypolicy-requirements
https://github.com/IBM/charts/tree/master/stable/ibm-mqadvanced-server-dev#podsecuritypolicy-requirements

[Required] Please change `ICP` reference to IBM Cloud Private in the Readme line#18 under Container Image Security

[Required] Passing secret values in a helm release
A single Tiller instance is used to store all Helm releases in a single location.
Anyone who has the authority to access Helm (an ICP Cluster Admin, Team Admin or Team Operator) has the ability to list the contents of any Helm release, regardless of the namespace that it was installed into.
Any cluster user, can run `helm get RELEASE` and get all the sensitive data that is passed to a release as values.
So sensitive data like license keys should be handled separately from the chart and release. The chart does provide an option for users to provide a pre-created, existing secret containing the license key.
Providing a pre-created secret is the secure and recommended approach to handle sensitive data. Please remove config which accepts license key as values.
You can add notes in the readme on how to generate a secret with the license key information.

[Recommend] All containers should run as a nonroot user. If the chart functionality requires root access then we need to document the reason why the chart needs to run as root and document the reason in the Readme as well.

[Recommend] Ideal secanario is where workload doesn't require any privileges, capabilities or root user access. Such workloads can safely run anywhere.
It's a best practice to explicitly declare the security context to avoid any assumptions or misconfigurations. A security context defines privilege and access control settings for a Pod or Container.
Here is an example code snippet of a pod spec defining its security context where it does not need any extra privileges or capabilities and can run as a non-root user.
```
spec:
  template:
    spec:
      hostNetwork: false
      hostPID: false
      hostIPC: false
      securityContext:
        runAsNonRoot: true
        runAsUser: < non-zero, numeric id >
      containers:
        securityContext:
          privileged: false
          readOnlyRootFilesystem: true
          allowPrivilegeEscalation: false
          runAsNonRoot: true
          runAsUser: < non-zero, numeric id >
          capabilities:
            drop:
            - ALL
```
Please let us know the security context can be applied to this chart that grants only required privileges.

[Required] Below values are used but not defined in values.yaml and values-metadata.yaml. Commented values do not appear in the configure page on ICP UI.
Please add them so that they are visible in ICP UI config page and users can set values for them.
- .Values.mancenter.image.pullSecrets
- .Values.mancenter.nodeSelector
- .Values.mancenter.resources
- .Values.mancenter.persistence.storageClass
- .Values.mancenter.service.clusterIP
- .Values.service.clusterIP
- .Values.image.pullSecrets
- .Values.nodeSelector
- .Values.resources

You can add suitable defaults or keep the values empty.

[Required] Please add helm test to test chart install.

[Required] Please add `management-center:3.12` image to ibm_cloud_pak/manifest.yaml so that it will be considered for VA scanning.

[Required] Please add node affinity so that the pods are scheduled on the workers with suitable and matching architecture. It should be added to all files (deployments, stateful sets, helm tests etc) where pods are scheduled.
affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: beta.kubernetes.io/arch
            operator: In
            values:
            - amd64

Accepted values for achitecture are: amd64, ppc64le, s390x

[Required] For a cloud pak, the images are stored in the local docker registry on the cluster.
With IBM Cloud Private, the icp-image-manager creates an ImagePullSecret named sa-<namespace> when an image is pushed to the ICP image registry.
Non-default service accounts must specify this secret to ensure access equivalent to installing with default ServiceAccount.

Example ServiceAccount with sa-<namespace> and optionally an user specified pull secret:
```
apiVersion: v1
kind: ServiceAccount
imagePullSecrets:
  - name: sa-{{ .Release.Namespace }}
```

[Required] `chart` and `heritage` are required fields under spec.template.metadata.labels. mancenter-deployment.yaml and statefulset.yaml are missing these required labels.

[Required] `serviceName` is a required field in StatefulSet and it should point to the right service (hazelcast service) since we have two services in the chart.
For reference:
```
Each Pod in a StatefulSet derives its hostname from the name of the StatefulSet and the ordinal of the Pod. The pattern for the constructed hostname is $(statefulset name)-$(ordinal). The example above will create three Pods named web-0,web-1,web-2. A StatefulSet can use a Headless Service to control the domain of its Pods. The domain managed by this Service takes the form: $(service name).$(namespace).svc.${cluster-domain}. As each Pod is created, it gets a matching DNS subdomain, taking the form: $(podname).$(governing service domain), where the governing service is defined by the serviceName field on the StatefulSet.
```
[Required] Similar to tillerVersion, please add `kubeVersion` to Chart.yaml
From the Readme, kubeVersion: ">1.9" looks like the right value.

[Required] Please add missing keyword `Commercial` in Chart.yaml keywords section. This keyword indicates that this chart will need a license key for deployment.

[Required] Please add `Support` section in the Readme. Example: https://github.com/IBM/charts/tree/master/community/postgresql#support

[Required] Please add RELEASENOTES.md
Example for this file will be: https://github.com/IBM/charts/blob/master/stable/ibm-mqadvanced-server-dev/RELEASENOTES.md
https://github.com/IBM/charts/blob/master/stable/ibm-websphere-liberty/RELEASENOTES.md

[Required] Apologize for this. Can you please add a `.` at tail end of qualification.terms in qualification.yaml.

[Required] Apologize for making this required.
In values-metadata.yaml, the description and label should not be identical.
I found 26 values which have identical description. Please change the description to be slightly more elaborate than the label.
I can also send you a list of the values if that will help.

Again excuse me for the long comment. I am testing a couple minor UI sugar for the chart. Will post comments if there are changes required. Please let me know if you have questions or need help.
Thank you.

mrgadgil commented:
@hasancelik @leszko 
Found a missing key `cluster.memberCount` in values-metadata.yaml.
Should be:
```
cluster:
  memberCount:
    __metadata:
      name: "memberCount"
      label: "Member Count"
      description: "StatefulSet memberCount"
      type: "number"
      immutable: false
      required: true
```

leszko commented:
@mrgadgil Thanks a lot for the good code review, I'll address the comments.

In the meantime, I have a problem with the ICP 3.1.2 installation (on AWS). What is the fastest way to install ICP on AWS? I tried Quick Start, but it fails. The other option is terraform or installation [from scratch](https://www.ibm.com/support/knowledgecenter/en/SSBS6K_3.1.2/installing/installing.html). What solution do you recommend?

Also, I'll probably need `ibm-cloud-private-x86_64-3.1.2.tar.gz`, how can I get it? The doc points to Passport Advantage, but I don't think we have an account there. Could you create one for us?

Thanks!

CC: @bgower-ibm

mrgadgil commented:
@leszko ICP 3.2 is the latest version of ICP and it GAed last Friday May 31st.
Cloud paks are required to be tested on latest ICP 3.2 environment so we recommend you to set up ICP 3.2 and test the chart on the latest and greatest.
AWS quickstart doesn’t support ICP 3.2 yet.
We are trying to get you IBM Cloud Private-CE (Community Edition) 3.2 images. I will update as soon we get more information.
For your reference : ICP CE 3.1.2 images are here: https://hub.docker.com/r/ibmcom/icp-inception

leszko commented:
@mrgadgil I addressed all your comments. Could you please have a look?

mrgadgil commented:
@leszko Thank you much! I will take a look at the code changes.
You will be able to find ICP 3.2 Community Edition images and installation guidelines here: https://hub.docker.com/r/ibmcom/icp-inception

mrgadgil commented:
@leszko 
Thank you for quick response and addressing my review comments.
I have added some comments directly on the files as I thought it would be easier to point code line numbers directly. Also below are few comments.

[Recommend] Security context
You have added security context to containers and pods so that they run as non-root user and have readonlyfilesystem.
I found that management-center deployment fails with `readOnlyRootFilesystem: true` 
I am running into write permission issues after disabling readonly file system setting **`readOnlyRootFilesystem: false`**

> 2019-06-04 21:20:59 [main] WARN  o.eclipse.jetty.webapp.WebAppContext - Failed startup of context o.e.j.w.WebAppContext@69cac930{Hazelcast Management Center,/hazelcast-mancenter,[file:///tmp/jetty-0.0.0.0-8080-hazelcast-mancenter-3.12.war-_hazelcast-mancenter-.WebApp-4411624122620844588.dir/webinf/, jar:file:///opt/hazelcast/mancenter/hazelcast-management-center-3.12/hazelcast-mancenter-3.12.war!/],UNAVAILABLE,@WebApp}{file:/opt/hazelcast/mancenter/hazelcast-management-center-3.12/hazelcast-mancenter-3.12.war}
> org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'sqlDbConfig' defined in URL [jar:file:/opt/hazelcast/mancenter/hazelcast-management-center-3.12/hazelcast-mancenter-3.12.war!/com/hazelcast/webmonitor/config/SqlDbConfig.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'homeDirectoryProviderImpl' defined in URL [jar:file:/opt/hazelcast/mancenter/hazelcast-management-center-3.12/hazelcast-mancenter-3.12.war!/com/hazelcast/webmonitor/service/HomeDirectoryProviderImpl.class]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.hazelcast.webmonitor.service.HomeDirectoryProviderImpl]: Constructor threw exception; nested exception is com.hazelcast.webmonitor.service.exception.WriteNotAllowedException: WARNING: /data can not be created. Either make it writable, or set "hazelcast.mc.home" system property to a writable directory and restart.

I am using a NFS persistence volume. I have give write permissions for group 1001 on the mount.
Note: The chart deploys when I disable persistence for management center.

Is it not mandatory to set the most restrictive security context. Please add one that works for the chart.
I think you will be able to debug more when you have a ICP 3.2 test environment. Please let me know if you have any questions. Will be happy to try and find answers.

[Required] Please add `alpine-bash-curl-ssl:0.3.0` image to ibm_cloud_pak/manifest.yaml so that it will be considered for VA scanning.

[Recommend] Thank you for adding helm test. Something to consider would be adding helm hook "helm.sh/hook-delete-policy" annotation.
Helm hooks are not automatically deleted by helm when release is removed. You can add hook delete policy for the hook to be deleted irrespective of helm test result.
`"helm.sh/hook-delete-policy": hook-succeeded, hook-failed`

[Required] Thank you for fixing keys in values-metadata.yaml where description and label should not identical.
I found 5 more cases where the description to be slightly more elaborate than the label.
livenessProbe.failureThreshold.__metadata
mancenter.image.pullPolicy.__metadata
mancenter.livenessProbe.enabled.__metadata
mancenter.livenessProbe.periodSeconds.__metadata
mancenter.readinessProbe.enabled.__metadata
mancenter.readinessProbe.periodSeconds.__metadata

[Required] Add metadata for cluster.__metadata in values-metadata.yaml.

Thank you.
Manjiree.

leszko commented:
@mrgadgil I've addressed your comments. PTAL.

Concerning the Management Center and `readOnlyRootFilesystem`, let's leave it `false` for now. I think that we would need to make changes in the Management Center application itself to change it to `true`.

Concerning the elaboration on the labels/descriptions in `values-metadata.yaml`. Please take a look. I've changed some of them. However, some (like `mancenter.readinessProbe.periodSeconds.__metadata`) should be ok. If you think I should change them, please write me more details what's wrong.

Thanks!
R.

mrgadgil commented:
@leszko 
[Required]
On investigating I found for cloud paks, the acceptance condition is description should be longer in length and not identical to label.
So can you please make the below changes to the description values-metadata.yaml
- mancenter.image.pullpolicy
description: "Allowed values Always, Never, or IfNotPresent."
- mancenter.readinessProbe.periodSeconds
description: "Interval of the periodic readiness probe checks for management center"
- mancenter.livenessProbe.periodSeconds
description: "Interval of the periodic liveness probe checks for management center"

Changes for you to consider while testing on ICP 3.2:
As I mentioned yesterday, I am running into write permissions while deploying management center.
To solve the issue, instead of using a fixed mountPath `/data` for volumeMount `mancenter-storage`, I changed it to a variable `.Values.mancenter.persistence.storage`
I then set the variable to the NFS mount which had write permissions for the fsgroup.
With above changes, I was able to deploy the chart on ICP 3.1.2 with persistence enabled and using a NFS volume and could run a sanity test.

Thanks!
Manjiree.

leszko commented:
@mrgadgil I've updated `values-metadata.yaml`. It should fulfill the requirements now.

Concerning `/data` and `mancenter-storage`, I don't think I follow. In your case, you created the NFS Storage manually, then you should just set `mancenter.persistence.existingClaim` to your storage name and that's all. It should work fine.
The default behavior is, however, that the Helm chart creates the persistent volume claim for you and mount it automatically into your container. And here, I'm not sure why, it is always mounted with the root permissions. Therefore, we need to have the following parameter: `readOnlyRootFilesystem: false`.
We can't change `/data` or extract it, since the volume needs to be mounted always into this directory. That's the specification inside the Management Center Dockerfile.

Cheers,
Rafał

leszko commented:
I've managed to install ICP 3.2.0 and tested this Helm Chart. It works correctly.

![image](https://user-images.githubusercontent.com/2834997/59023087-50bf0280-884f-11e9-88e5-b50c4621e2e9.png)

mrgadgil commented:
@leszko 
I got the chart install to work with NFS mount as well. My bad. After I gave right permissions to the NFS mount, I am able to test chart install. I had given the correct group permissions but missed setting userid.

I have another review comment which I missed sharing yesterday.
NOTES.txt also should be updated to use the new labels. 
`app` should be changed to `app.kubernetes.io/name`
Please verify instructions in NOTES.txt making sure that they work for every service type.

leszko commented:
@mrgadgil Good point about `NOTES.txt`. I updated the labels + updated the text a little bit. 
Also I tested it all together for all service types.

PTAL.

mrgadgil commented:
@leszko 
I have another recommendation for access mode for PVC.
It is disadvantageous for Deployments to point to PVCs with an access mode of ReadWriteOnce because multiple Pods cannot concurrently access the same ReadWriteOnce storage. 
I understand there would be only 1 replica for the management center deployment. But still it is recommended not to use RWO access mode.
https://cloud.google.com/kubernetes-engine/docs/concepts/persistent-volumes#deployments_vs_statefulsets
Quoting from the article above:

> Even Deployments with one replica using a ReadWriteOnce Volume are not recommended. This is because the default Deployment strategy will create a second Pod before bringing down the first pod on a recreate. The Deployment may fail in deadlock as the second Pod can't start because the ReadWriteOnce Volume is already in use, and the first Pod wont be removed because the second Pod has not yet started.Instead, use a StatefulSet with ReadWriteOnce volumes.

My recommendation would be to change default access mode for management center to `ReadWriteMany`.

leszko commented:
@mrgadgil This is a very valid comment! Thanks for it. Actually, the solution is not to change accessMode to `ReadWriteMany`, but rather change Management Center from `Deployment` to `StatefulSet`. Management Center will fail if two instances target the same directory, because of the lock file inside. Management Center is an app which is not prepared to run in multiple instances. So, according to the documentation you quoted, it should be a StatefulSet.

I implemented and tested the change. Please have a look.

mrgadgil commented:
@leszko Perfect! I was not sure if you do a design change.
Yes I did run into issues when I tried to do multiple installs of the chart but I was using the same NFS mount directory.
Error that I got was:
```
Caused by: java.lang.IllegalStateException: The file is locked: nio:/data/sql/mc.mv.db [1.4.197/7]
```
That is why, I thought management center is supposed to have only 1 replica. Stateful is the recommended way of deploying stateful applications. I will take a look at the change.

leszko commented:
@mrgadgil I addressed the Vulnerability Advisor comments, please take a look.

mrgadgil commented:
@leszko I tested `helm test` with the updated images. Looks good.
Found a typo in ReadMe. Please correct when you get a chance.
Thanks much!
Manjiree.

mrgadgil commented on community/hazelcast-enterprise/README.md:
````diff
@@ -29,12 +23,79 @@ spec:
+You need to create a secret with the Hazelcast Enterprise License Key:
+
+```bash
+$ kubectl create secret generic hz-enterprise-license-key --from-literal=key=LICENSE-KEY
````
Please remove the extra `=` between `literal` and `key`
--from-literal=key=LICENSE-KEY ===> --from-literal key=LICENSE-KEY

leszko replied:
`kubectl create secret generic hz-enterprise-license-key --from-literal=key=LICENSE-KEY` is actually a correct syntax. You can find similar usages, for example, here: https://github.com/IBM/charts/blob/master/community/artifactory-ha/README.md

mrgadgil replied:
yes it is indeed. Please ignore my comment. Looks good. Thanks.

leszko commented:
@mrgadgil I replied to you comment. PTAL.

mrgadgil commented:
@leszko Sorry for the delayed response. Hopefully you must have received an email from Scott Anderson about changes in direction.
Please remove the ibm_cloud_pak directory. I will approve the PR after the change.

leszko commented:
@mrgadgil I've updated `qualification.yaml`. PTAL.

mrgadgil commented:
@leszko Super apologies for requesting another change.
I missed how the chart will look on the ICP configure UI when I asked you to blank only the two keys levelName and levelDescription from the qualification.yaml
When I tested loading the chart, it gave me a missing value on the ICP configure UI.
![image](https://user-images.githubusercontent.com/49280244/59634974-97491280-9115-11e9-9ca7-3595b16c875a.png)

The right solution would be to get rid of `qualification` and all 5 subkeys in the qualification.yaml.
```
prereqs:
  security:
    kubernetes:
      podSecurityPolicy:
        name: "ibm-restricted-psp"
    ibmCloudPrivate:
      installerRole:
        name: "Operator"
```
This should be final final change. I will approve the PR and proceed with the merge.
You might already know this. 
You can load the chart and see how the chart config/readme looks in the UI with below steps:
- Obtain and install cloudctl from https://<CLUSTER_MASTER_IP>:8443/console/tools/cli
- Login to the cluster using cloudctl
- helm package hazelcast-enterprise/
- cloudctl catalog load-chart --archive hazelcast-enterprise-1.0.3.tgz --repo <HELM_REPO>
- Click on catalog and select the chart
After fix, UI is rendered the right way:
![image](https://user-images.githubusercontent.com/49280244/59635583-34f11180-9117-11e9-8cf7-782109664bc8.png)

leszko commented:
@mrgadgil Updated. PTAL.

mrgadgil commented:
@leszko 
We are in the processing of finalizing the revised contribution guidelines for charts added in the catalog. Our guidelines document will soon reflect updated criteria.
I was doing additional testing for the chart on ICP 3.1.1 and found that getting rid of qualification and all 5 subkeys in the qualification.yaml generates errors in displaying the catalog UI. 
Note: This is a bug in ICP 3.1.1 catalog where catalog ui blows up on not finding qualification and its sub-keys. This is fixed in ICP 3.1.2 and ICP 3.2
Having said that, the chart has to support and work on 3.1 (+ all patch versions) and 3.2 because it is part of community charts.
So I would request another change to qualification yaml.
```
qualification:
  levelName: ""
  levelDescription: ""
  issueDate: ""
  duration: ""
  terms: ""
prereqs:
  security:
    kubernetes:
      podSecurityPolicy:
        name: "ibm-restricted-psp"
    ibmCloudPrivate:
      installerRole:
        name: "Operator"
```
Sorry for the confusion.
Thank you
Manjiree.

leszko commented:
@mrgadgil , updated. PTAL.

mrgadgil commented:
@leszko Found a mention to cloud pak certification.
Need to remove reference to `Cloud Pak certification` from RELEASENOTES.md since the term cloud pak has different meaning now.
Barry will merge after the above is addressed. Thank you!

leszko commented:
@mrgadgil @bgower-ibm Updated!
